reco_utils/recommender/ripplenet/model.py

```diff
@@ -0,0 +1,241 @@
+
+            # [batch_size, dim, 1]
+            v = tf.expand_dims(self.item_embeddings, axis=2)
+
+            # [batch_size, n_memory]
+            probs = tf.squeeze(tf.matmul(Rh, v), axis=2)
+
+            # [batch_size, n_memory]
+            probs_normalized = tf.nn.softmax(probs)
+
+            # [batch_size, n_memory, 1]
+            probs_expanded = tf.expand_dims(probs_normalized, axis=2)
+
+            # [batch_size, dim]
+            o = tf.reduce_sum(self.t_emb_list[hop] * probs_expanded, axis=1)
+
+            self.item_embeddings = self.update_item_embedding(self.item_embeddings, o)
+            o_list.append(o)
+        return o_list
+
+    def update_item_embedding(self, item_embeddings, o):
+        if self.item_update_mode == "replace":
+            item_embeddings = o
+        elif self.item_update_mode == "plus":
+            item_embeddings = item_embeddings + o
+        elif self.item_update_mode == "replace_transform":
+            item_embeddings = tf.matmul(o, self.transform_matrix)
+        elif self.item_update_mode == "plus_transform":
+            item_embeddings = tf.matmul(item_embeddings + o, self.transform_matrix)
+        else:
+            raise Exception("Unknown item updating mode: " + self.item_update_mode)
+        return item_embeddings
+
+    def predict_scores(self, item_embeddings, o_list):
+        y = o_list[-1]
+        if self.using_all_hops:
+            for i in range(self.n_hop - 1):
+                y += o_list[i]
+
+        # [batch_size]
+        scores = tf.reduce_sum(item_embeddings * y, axis=1)
+        return scores
+
+    def _build_loss(self):
+        self.base_loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=self.labels, logits=self.scores))
+
+        self.kge_loss = 0
+        for hop in range(self.n_hop):
+            h_expanded = tf.expand_dims(self.h_emb_list[hop], axis=2)
+            t_expanded = tf.expand_dims(self.t_emb_list[hop], axis=3)
+            hRt = tf.squeeze(tf.matmul(tf.matmul(h_expanded, self.r_emb_list[hop]), t_expanded))
+            self.kge_loss += tf.reduce_mean(tf.sigmoid(hRt))
+        self.kge_loss = -self.kge_weight * self.kge_loss
+
+        self.l2_loss = 0
+        for hop in range(self.n_hop):
+            self.l2_loss += tf.reduce_mean(tf.reduce_sum(self.h_emb_list[hop] * self.h_emb_list[hop]))
+            self.l2_loss += tf.reduce_mean(tf.reduce_sum(self.t_emb_list[hop] * self.t_emb_list[hop]))
+            self.l2_loss += tf.reduce_mean(tf.reduce_sum(self.r_emb_list[hop] * self.r_emb_list[hop]))
+            if self.item_update_mode == "replace nonlinear" or self.item_update_mode == "plus nonlinear":
+                self.l2_loss += tf.nn.l2_loss(self.transform_matrix)
+        self.l2_loss = self.l2_weight * self.l2_loss
+
+        self.loss = self.base_loss + self.kge_loss + self.l2_loss
+
+    def _build_train(self):
+        self.optimizer = tf.train.AdamOptimizer(self.lr).minimize(self.loss)
+        '''
```

Comment:
I'm a fan of testing different optimization methods. Would it be possible to do something to select the optimizer? an example can be seen here: https://github.com/microsoft/recommenders/blob/master/reco_utils/recommender/deeprec/models/base_model.py#L232 (this example in fact would be more pythonic if instead of we use chained ifs, we use dictionaries, like here https://github.com/microsoft/recommenders/blob/master/benchmarks/benchmark_utils.py#L318)

Reply:
I tried to implement this change with a dictionary, but the code was failing for some internal tensorflow error, I think it's because it tries to implement all optimizers. Could you take a look? f1e590dbbdafe1c6bdc2193888dc2180d9346954

Reply:
might be because it is needed to add the inputs externally, see some examples https://github.com/microsoft/recommenders/blob/master/benchmarks/movielens.ipynb. But, when it is a little bit complicated, ifs can be more readable. Which way do you think it is clearer for the user?

Reply:
I think in this case the if-else list is not too long, and it has the same structure as __update_item_embedding_